### ISSUE TITLE
Playing music causes Python to use all CPU memory

### ISSUE THREAD

**Issue by anthonyrgib** [Cog-Creators/Red-DiscordBot#946]
Hello,

I will try and explain my issue in as great detail as possible. I've owned a Red bot, largely without issue for over 7 months on the same VPS server (important later on).

Recently, about 2 days ago I did an update as I saw this in the Red Discord server announcements: "Updating the requirements from the launcher should fix the status issue". So as usual, I updated pip, updated Red and updated requirements. I then proceeded to install basic requirements.

Everything went fine, the bot started up as normal and worked absolutely perfectly.

Users notified me that there was lag when music was playing, I started trying to figure out what was causing lag and I noticed the CPU usage for Python (It's a windows server VPS) was bouncing around, from 1% CPU to 80% to 45% and so on and so forth.

I spoke to some of the help staff and other members, updated all cogs, updated the audio requirements, tried everything. 

I've tried changing the music format from FFMPEG to AVCON (I may misspell these as i'm not familiar with them).

I've also unloaded ALL cogs, even the core ones and restarted the bot, loaded only Audio, tried to play music, and again, the CPU memory starts to flutter and go all over the place.

I hope i've given enough detail, I will keep an eye on this thread and provide additional details if required.

Thanks in advance.

**Comment by palmtree5** [comment]
I'm wondering: what's the CPU usage with the bot not running?

**Comment by DannyB5544** [comment]
From my CPU graphs, I barely hit 5% unless playing music, where it is a little sparadic. But not as bad as this.

**Comment by anthonyrgib** [comment]
It's pretty stable right now with the CPU at 5% CPU usage, Python is only using about 2.5% CPU usage, it's only when music plays does Python spike, but it's stable without music running.

Some images from yesterday of the Task Manager when music is playing:

https://prntscr.com/gcdo73
https://prntscr.com/gcdftz
https://prntscr.com/gcd2mf

**Comment by DannyB5544** [comment]
I'll give the audio commands a poke later, see if I can recreate the issue.

**Comment by palmtree5** [comment]
TBH I don't have the issue of Python's CPU usage spiking when playing audio. Then again, my instance is running on a single-core VPS (with less RAM than what your VPS has according to a screenshot you posted in the support channel last night) running Ubuntu 16.04. So I'm wondering at this point if this is an issue that's restricted to Windows

**Comment by DannyB5544** [comment]
That would make sense. Since I switched to Linux I stopped getting it.

**Comment by Twentysix26** [comment]
How many servers are running audio at the same time?

**Comment by anthonyrgib** [comment]
The bot is only on one server, so it's just one

**Comment by palmtree5** [comment]
What else, if anything, are you running on this VPS that a) is at all CPU-intensive, and b) requires you to use Windows?

**Comment by anthonyrgib** [comment]
I have no idea how to use linux which is why i use windows server, and nothing else runs on the VPS, I can take another screenshot of Task Manager if you like

**Comment by anthonyrgib** [comment]
Another bot i'm running was just having music issues, and a .dll file was missing from the server. Fixed it now, but because of that youtube-dl wasn't working properly. That bot is fixed now but Nadeko is still lagging when playing music (python going crazy CPU usage wise). Is it possible that my version of youtube-dl that Red is using is broken or corrupted? All I can think is, whatever the issue it has to be isolated to just me.

Is there any way to like delete the youtube-dl folder in /lib and it will reinstall itself?

**Comment by Stylite** [comment]
This is not nadeko. pip uninstall youtube_dl would uninstall yt-dl and reinstall it using the launcher.

**Comment by anthonyrgib** [comment]
Sorry, I meant to write Red not nadeko, that's the other bot that had the issue. My bad. 

So I need to wipe the lib folder in Red's launcher and then update pip and requirements etc? I can't find any pip uninstall

**Comment by Kowlin** [comment]
I've been thinking about this for some time now. And it's absolutely possible that Red doesn't use proper settings for YouTube DL that optimise for CPU in mind. For now I don't have any plans to fix this. But I will be looking into it with the guys who do audio for v3.

**Comment by anthonyrgib** [comment]
Today I tried to troubleshoot on my own, I tried wiping the /lib folder using the launcher and reinstalling basic and audio requirements, these had no effect. I then tried using previous versions of youtube-dl and other folders found in the /lib folder. Again, to no avail. 

I then tried removing the audio.py file, but red will not automatically redownload this file if it appears to be missing. I placed it back in, obviously there were no changes in CPU usage.

I then decided to uninstall Python completely and reinstall it. Again, to no avail. 

Lastly, I then decided to just reinstall my VPS, wiping it clean completely and giving me a fresh start. This seems to, at the moment at least, fix the issue. So it may be the case that in my isolated instance, it just was an issue with the VPS or some windows service causing Python to fluctuate when playing audio. 

I'm very grateful that the Red developers, staff and other members of the Red community looked so hard at my issue and tried their best to help me.

Another bot i'm using, Nadeko, which also uses youtube-dl doesn't have these issues, so your message about reviewing the settings to optimise for Youtube-dl is very welcoming.

While searching the web for any clues, all i could stumble on were unrelated threads where Python is using high memory being cited as a script issue. I was not able to isolate which script specifically was causing the CPU to fluctuate so but it may be an area of investigation for future optimisation.

For me personally, the issue seems to have been largely resolved by resetting my VPS and I'm happy that my specific issue is closed. 

Thanks to everyone who replied and helped me out. Red definitely has the best support system.
